ComplexTextController: Control characters should be rendered as visible glyphs

fujii opened the pull request:
#### 3baa72dbb8d513476b27bcfae8d2dd02c6514541
<pre>
ComplexTextController: Control characters should be rendered as visible glyphs
<a href="https://bugs.webkit.org/show_bug.cgi?id=271869">https://bugs.webkit.org/show_bug.cgi?id=271869</a>

Reviewed by Vitor Roriz.

Control characters weren&apos;t rendered as visible glyphs in the complex
text code path in some condition.

The spec &lt;<a href="https://www.w3.org/TR/css-text-3/#white-space-processing">https://www.w3.org/TR/css-text-3/#white-space-processing</a>&gt; says:

&gt; Control characters (Unicode category Cc)—other than tabs (U+0009),
&gt; line feeds (U+000A), carriage returns (U+000D) and sequences that form
&gt; a segment break—must be rendered as a visible glyph

<a href="https://webkit.org/b/149128">https://webkit.org/b/149128</a> fixed the simple text code path.
<a href="https://webkit.org/b/153941">https://webkit.org/b/153941</a> fixed a part of the complex text code
path. We need one more fix.

ComplexTextController constructor is using
FontCascade::isCharacterWhoseGlyphsShouldBeDeletedForTextRendering to
determine a character is visible. It returned false for control
characters.

Changed isCharacterWhoseGlyphsShouldBeDeletedForTextRendering to
return false for control characters other than tabs, line feeds, and
carriage returns.

WidthIterator::applyCSSVisibilityRules is also using
isCharacterWhoseGlyphsShouldBeDeletedForTextRendering. Changed it to
keep the original behavior.

* LayoutTests/fast/text/control-characters/complex-1-expected.txt: Added.
* LayoutTests/fast/text/control-characters/complex-1.html: Added.
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::isCharacterWhoseGlyphsShouldBeDeletedForTextRendering):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::applyCSSVisibilityRules):

Canonical link: <a href="https://commits.webkit.org/277574@main">https://commits.webkit.org/277574@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ee1c464dcf42fe7b5fbb21d03a54654e68143fd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47917 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27118 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50775 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50598 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43970 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50223 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32990 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24605 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38983 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24786 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41432 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20281 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22261 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42615 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5964 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44263 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43053 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52492 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22959 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19311 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46292 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24229 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41570 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45336 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10593 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25020 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23950 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on LayoutTests/fast/text/control-characters/complex-1.html:
I think we should make this a WPT test, specially because we are making web-combat choices.

fujii replied:
WPT already has a test. But, WebKit takes the simple text code path for the test. Deciding how to choose a  code path is WebKit specific. That's the reason why I hesitate to add a test to wpt.
https://wpt.live/css/css-text/white-space/control-chars-001.html

But, I will try if you still think it should be a wpt test.

vitorroriz replied:
Right, that makes sense. However, this can still be a WPT test as other browsers will render it the same and specially because we are guarding the behavior:

` Also, we're omitting Null (U+0000) from this set because Chrome and Firefox do so and it's needed for compat. See https://github.com/w3c/csswg-drafts/pull/6983.`

fujii replied:
Because Myles's request was rejected, all browsers don't conform to CSS spec.
https://github.com/w3c/csswg-drafts/pull/6983#issuecomment-1366336170
https://bug-271869-attachments.webkit.org/attachment.cgi?id=470687

If I add a test case of Null (U+0000), WebKit won't pass the test.
Do you mean adding a test which doesn't comform to the CSS spec?

vitorroriz replied:
I mean, if Blink, Gecko and now WebKit will behave in a certain way and this way differs what spec says, that would be a good reason to open a spec issue. I've seen WPT tests differing from spec before, although I agree that it might not be always the recommended thing to do. We could make a WPT and open a spec issue, or we could open a spec issue first.

vitorroriz replied:
That said, no reason to block the PR because of that ofc.

fujii replied:
Hmm, IMHO, Myles opened https://github.com/w3c/csswg-drafts/pull/6983 for the reason, and rejected. Nothing has changed since then.

To complicate matters, all browsers suprisingly pass the test case control-chars-000.html.
https://wpt.fyi/results/css/css-text/white-space/control-chars-000.html?label=master&label=experimental&aligned

Handling Null (U+0000) is done by multiple layers.
Actually, the test content has a tricky hack.
https://github.com/web-platform-tests/wpt/blob/c1247636413abebe66ca11a2ca3476de771c99cb/css/css-text/white-space/control-chars-000.html#L13

This patch actually fixes a WebKit bug. https://webkit.org/b/271869
I'd like to land this patch if the implementation has no problem.

vitorroriz commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -802,10 +802,16 @@ ComplexTextController::ComplexTextRun::ComplexTextRun(const Font& font, const UC
+        // https://www.w3.org/TR/css-text-3/#white-space-processing
+        // "Control characters (Unicode category Cc)—other than tabs (U+0009), line feeds (U+000A), carriage returns (U+000D) and sequences that form a segment break—must be rendered as a visible glyph"
+        // Also, we're omitting Null (U+0000) from this set because Chrome and Firefox do so and it's needed for compat. See https://github.com/w3c/csswg-drafts/pull/6983.
+        if (character == nullCharacter || character == tabCharacter || character == newlineCharacter || character == carriageReturn)
```
I wonder if this shouldn't be part of `isCharacterWhoseGlyphsShouldBeDeletedForTextRendering ` instead. You have mentioned that this is already fixed for WidthIterator, which also uses this function, so maybe embedding this logic into `isCharacterWhoseGlyphsShouldBeDeletedForTextRendering` would allow us to eliminate some code on WidthIterator also.

fujii replied:
Fixed. Could you take a look again?

fujii committed:
ComplexTextController: Control characters should be rendered as visible glyphs
https://bugs.webkit.org/show_bug.cgi?id=271869

Reviewed by Vitor Roriz.

Control characters weren't rendered as visible glyphs in the complex
text code path in some condition.

The spec <https://www.w3.org/TR/css-text-3/#white-space-processing> says:

> Control characters (Unicode category Cc)—other than tabs (U+0009),
> line feeds (U+000A), carriage returns (U+000D) and sequences that form
> a segment break—must be rendered as a visible glyph

https://webkit.org/b/149128 fixed the simple text code path.
https://webkit.org/b/153941 fixed a part of the complex text code
path. We need one more fix.

ComplexTextController constructor is using
FontCascade::isCharacterWhoseGlyphsShouldBeDeletedForTextRendering to
determine a character is visible. It returned false for control
characters.

Changed isCharacterWhoseGlyphsShouldBeDeletedForTextRendering to
return false for control characters other than tabs, line feeds, and
carriage returns.

WidthIterator::applyCSSVisibilityRules is also using
isCharacterWhoseGlyphsShouldBeDeletedForTextRendering. Changed it to
keep the original behavior.

* LayoutTests/fast/text/control-characters/complex-1-expected.txt: Added.
* LayoutTests/fast/text/control-characters/complex-1.html: Added.
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::isCharacterWhoseGlyphsShouldBeDeletedForTextRendering):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::applyCSSVisibilityRules):

Canonical link: https://commits.webkit.org/277574@main
